### DISCUSSION TITLE
TS failure on import

### DISCUSSION THREAD

**Discussion by DawidJanczak** [orchidjs/tom-select#100] [Q&A]
Hello!

Possibly related to https://github.com/orchidjs/tom-select/issues/66, we're having some problems importing this in a webpack project set up with TS. Having a `import TomSelect from 'tom-select' ` and then running `npx tsc` results in the following error:
```
index.ts:1:23 - error TS7016: Could not find a declaration file for module 'tom-select'. '/var/app/node_modules/tom-select/dist/js/tom-select.complete.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/tom-select` if it exists or add a new declaration (.d.ts) file containing `declare module 'tom-select';`

1 import TomSelect from 'tom-select'
                        ~~~~~~~~~~~~


Found 1 error.
```

Any idea what's wrong here?

**Comment by oyejorge** [comment]
Strange. TypeScript is attempting to import TomSelect from ```/var/app/node_modules/tom-select/dist/js/tom-select.complete.js``` (which is not an ESM module) even though the module path is defined as ```dist/esm/tom-select.complete.js``` in the package.json.

**Reply by oyejorge** [reply]
Does using this import statement fix the problem?

```import TomSelect from 'tom-select/dist/esm/tom-select.complete.js'```

**Reply by DawidJanczak** [reply]
Unfortunately not, that results in
```
index.ts:1:23 - error TS7016: Could not find a declaration file for module 'tom-select/dist/esm/tom-select.complete.js'. '/var/app/node_modules/tom-select/dist/esm/tom-select.complete.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/tom-select` if it exists or add a new declaration (.d.ts) file containing `declare module 'tom-select/dist/esm/tom-select.complete.js';`

1 import TomSelect from 'tom-select/dist/esm/tom-select.complete.js'
                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 1 error.
```

This is just a shot in the dark, but I noticed that `types` field in `package.json` points to a directory and in examples on https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html#including-declarations-in-your-npm-package they always point to a file. Might this have something to do with it?

**Reply by oyejorge** [reply]
Success!... sort of. I am getting the error now. But only when I run ```npx tsc --strict true```

> but I noticed that types field in package.json points to a directory ... Might this have something to do with it?

Yes! Changing that to ```"types": "src/tom-select.complete.ts",``` solves the problem you're seeing.

However, with ```--strict true``` there are a lot of other errors that pop up. The package still needs some work for strict type checking to pass.

**Reply by DawidJanczak** [reply]
Ah good spot! We do have `strict` in our TSConfig indeed. I thought I tested this without TS options just to be sure it's nothing in our config, but it seems I did not do that properly, apologies. `strict` option is a "catch-all" for a few other options and I narrowed it down to these two: https://www.typescriptlang.org/tsconfig#alwaysStrict and https://www.typescriptlang.org/tsconfig#noImplicitAny All other strict checks mentioned under https://www.typescriptlang.org/tsconfig#strict are passing.

**Reply by oyejorge** [reply]
Made some more progress on typings in 1.7.3 but there's still quite a few noImplicitAny issues. Gonna keep plugging away...

**Reply by DawidJanczak** [reply]
Let me know if you need a hand with that. JS/TS is not my speciality by any means, but I should be able to help a bit.

**Reply by oyejorge** [reply]
Yeah, would be more than happy if you helped me fix some typing issues!

**Reply by DawidJanczak** [reply]
Sure thing, I should have some time over the weekend. Any particular file(s) I should start on? Wouldn't want to conflict with your changes if you're planning any. If not I'll just pick something at random.

**Reply by oyejorge** [reply]
Awesome, thanks! It might make sense to start with utils.ts and vanilla.ts, but I don't think it really matters.

**Reply by charlesmudy** [reply]
Not sure if there is a solution around this? Having same issue
